test/tstDetailsHalfTraversal.cpp

```diff
@@ -0,0 +1,106 @@
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborXTest_StdVectorToKokkosView.hpp"
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsHalfTraversal.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include <Kokkos_Random.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+namespace Test
+{
+template <class ExecutionSpace>
+// enclosing parent function for an extended __host__ __device__ lambda must not
+// have deduced return type
+#ifndef KOKKOS_COMPILER_NVCC
+auto
+#else
+Kokkos::View<ArborX::Point *, ExecutionSpace>
+#endif
+    make_random_cloud(ExecutionSpace const &space, int n)
+{
+  Kokkos::View<ArborX::Point *, ExecutionSpace> points(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing, "Test::points"),
+      n);
+  using RandomPool = Kokkos::Random_XorShift64_Pool<ExecutionSpace>;
+  RandomPool random_pool(5374857);
+
+  Kokkos::parallel_for(
+      "Test::generate_random_points",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int i) {
+        typename RandomPool::generator_type generator = random_pool.get_state();
+        auto const x = generator.frand(0.f, 1.f);
+        auto const y = generator.frand(0.f, 1.f);
+        auto const z = generator.frand(0.f, 1.f);
+        points(i) = {x, y, z};
+        random_pool.free_state(generator);
+      });
+
+  return points;
+}
+
+// Workaround NVCC's extended lambda restrictions
+struct AlwaysTrue
+{
+  template <class ValueType>
+  KOKKOS_FUNCTION bool operator()(ValueType const &) const
+  {
+    return true;
+  }
+};
+
+struct PredicateGetter
+{
+  template <class ValueType>
+  KOKKOS_FUNCTION AlwaysTrue operator()(ValueType) const
+  {
+    return {};
+  }
+};
+} // namespace Test
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(half_traversal, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  using MemorySpace = typename DeviceType::memory_space;
+  using ExecutionSpace = typename DeviceType::execution_space;
+  ExecutionSpace exec_space;
+  int const n = 24;
+  auto points = Test::make_random_cloud(exec_space, n);
+  ArborX::BVH<MemorySpace> bvh(exec_space, points);
+
+  Kokkos::View<int *, MemorySpace> count("Test::count", n * (n + 1) / 2);
+
+  using ArborX::Details::HalfTraversal;
+  HalfTraversal(
+      exec_space, bvh, Test::PredicateGetter{}, KOKKOS_LAMBDA(int i, int j) {
+#if KOKKOS_VERSION >= 30700
+        auto [min_ij, max_ij] = Kokkos::minmax(i, j);
+#else
+        auto [min_ij, max_ij] = Kokkos::Experimental::minmax(i, j);
+#endif
+        Kokkos::atomic_increment(&count(max_ij * (max_ij + 1) / 2 + min_ij));
```

Comment:
This needs an explanation of what you are trying to do. I struggle figuring it out.